Added the argmax and argmin in sympy.calculus

mijo2 opened the pull request:
#### References to other Issues or PRs
Fixes #13432 \
Closes #13555 

#### Brief description of what is fixed or changed

--> Added argmax and argmin functions in util.py 

#### Other comments

Example of usage:

```
>>> from sympy.abc import *
>>> from sympy.calculus.util import *
>>> from sympy import *
>>> argmin(sin(x), x, Interval(-2*pi, pi))
FiniteSet(-pi/2)
>>> argmax(sin(x), x, Interval(-2*pi, pi))
FiniteSet(-3*pi/2, pi/2)
```

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* calculus
  * Added argmax and argmin functions 
<!-- END RELEASE NOTES -->

mijo2 committed:
calculus: Added the argmax and argmin functionalities

Introduced in a fix but hasn't been resolved yet.

mijo2 commented:
I have tried to address this [comment](https://github.com/sympy/sympy/pull/13555#issuecomment-533868905) in this PR. But I am still not sure what to name the argmax and argmin functions. If someone can suggest some names then it would be great!

mijo2 committed:
calculus: Removed bare except clause

Bare except clause is usually not preferred and it also was the
reason why the build for the first commit of this PR failed. Hence,
in this commit, bare except was removed.

mijo2 committed:
calculus: Removing the unnecessary imports

oscarbenjamin commented:
Does the function work over non-real domains? If it is only intended for real domains it could be called e.g. `argmin_real`.

mijo2 commented:
@oscarbenjamin I have to check if it works in complex domain or not. If it doesn't, which is most likely the case then I will used the ```argmin_real``` and ```argmax_real``` as the names of the functions.

oscarbenjamin commented:
I wonder what is the best way to handle something like this:
```
In [17]: argmin((x - sqrt(2))**2, x, Rationals)                                                                                                
---------------------------------------------------------------------------
NotImplementedError
```

mijo2 committed:
calculus: Updated the docstring

Some updates were made in the docstring of argmax and argmin helper
functions as they were not in the right format and this was the main
reason the build failed.

mijo2 commented:
@oscarbenjamin I tried running ```argmin(x**2 - 2*x + 1, domain = S.Complexes)``` and as expected, it gave ```NotImplementedError```. So, I will rename the ```argmax``` and ```argmin``` functions to ```argmax_real``` and ```argmin_real```. I am thinking of also adding a check to verify whether a domain is a subset of ```Real``` or not and if it isn't, the function can raise a ```ValueError```.

mijo2 commented:
As for [this](https://github.com/sympy/sympy/pull/18550#issuecomment-581356285) comment, I don't know for sure as the function range function itself is unable to compute the range so we are not getting any extremum.

oscarbenjamin commented:
> I tried running `argmin(x**2 - 2*x + 1, domain = S.Complexes)` and as expected, it gave `NotImplementedError`

It is reasonable for this to give an error because the expression is not real valued over the domain so minimisation is not well defined. We need a function `f: C -> R` e.g.:
```
argmin(Abs(x), x, Complexes)
```

mijo2 commented:
@oscarbenjamin With ```argmin(Abs(x), x, Complexes)``` a ```ValueError``` is generated.

```
>>> argmin(Abs(x), x, Complexes)
Traceback (most recent call last):
...
ValueError: 
Absolute values cannot be inverted in the complex domain.
```

Also, with ```abs``` function, when I ran this: ``` argmin(Abs(x), x, S.Reals)```, a ```ValueError``` was generated again. 

```
>>> argmin(Abs(x), x, S.Reals)
Traceback (most recent call last):
...
ValueError: 
Can't calculate derivative wrt 0.
```
So, this function is able to compute argmax and argmin only for functions which are differentiable in the domain specified.

oscarbenjamin commented:
> So, this function is able to compute argmax and argmin only for functions which are differentiable in the domain specified.

Okay but what about `abs(x)**2`? That's everywhere differentiable for real x at least.

mijo2 commented:
@oscarbenjamin with ```abs(x)**2``` its still giving the same ```ValueError``` probably because it is unable to realise that ```abs(x)**2``` is equivalent to ```x**2``` and thus tries to find derivative and fail.

asmeurer commented:
There's also the question of functions with nonreal range, like sqrt(x) over R. Should argmax just ignore nonreal values or should it give an error?

mijo2 commented:
I think this depends on function range function defined in ```util.py``` as it is this function that does this kind of computation.

mijo2 commented:
@asmeurer With functions of nonreal range like ```sqrt(x)```, ```argmin```returns ```EmptySet``` because even though it is able to identify the infimum of the function, since derivative ```sqrt(x)``` at ```x=0``` becomes ```oo``` but not ```0```, so it doesn't register  ```x=0``` in critical points and thus isn't able to recognise ```x = 0``` is argmin

mijo2 committed:
calculus: Updated the function names edited

argmax and armin were named to argmax_real and argmin_real
appropriately.

mijo2 commented:
@oscarbenjamin @asmeurer I have changed the names to ```argmax_real``` and ```argmin_real```since that makes more sense given the fact that this works only in real domain. Also, in the docstring of this function it is mentioned that this function delivers correct results only when the function is differentiable in the given domain. Also, @oscarbenjamin regarding your comment about ```abs(x) ** 2```, I think the problem here maybe because of the fact that the functions that use to compute the function range aren't able to recognise that ```abs(x) ** 2``` is actually equivalent to ```x ** 2```, so I don't know how to correctly fix that issue yet.

oscarbenjamin commented:
You've added a top-level file called t.py and then deleted it again. Having that in the history can be problematic in the future e.g. when doing git bisect. Can you squash the commits?

mijo2 commented:
@oscarbenjamin Yes I made a mistake. I will squash the commits.

mijo2 commented:
@oscarbenjamin I have squashed the commits as you asked.

asmeurer commented:
I think just `argmin` and `argmax` are better. There's no reason why support for non-real domains couldn't be added in the future.

mijo2 commented:
@asmeurer Ok, I will rename them as ```argmax``` and ```argmin``` again. But after that, what should be the next step to improve this functionality?

oscarbenjamin commented:
I think that the next step would be making it possible to handle non-real domains and also finding minima that are not differentiable.

vezeli commented on sympy/calculus/util.py:
```diff
@@ -1597,3 +1597,155 @@ def union(self, other):
+
+def _argMaxMin(f, symbol, domain, max):
+    """
+    Helper function for calculuating argmin/argmax
```
Should these maybe be `argmin_real` and `argmax_real`?

mijo2 replied:
Well @asmeurer said [here](https://github.com/sympy/sympy/pull/18550#issuecomment-582744068) that we will add support for non-real domains in future so the names can be ```argmax``` and ```argmin```.

mijo2 commented:
@oscarbenjamin Can we check the points where the function is not differentiable and then add it to ```critical_points``` for checking if it is extremum or not?

oscarbenjamin commented:
I don't know if there is such a function. It would be useful both for real and complex functions to have something like that. There is singularities:
```
In [12]: from sympy.calculus import singularities                                                                                              

In [13]: singularities(1/(x**2-x), x)                                                                                                          
Out[13]: {0, 1}
```
That only works for rational functions though. Working over the reals we can rewrite as Piecewise:
```
In [16]: x = Symbol('x', real=True)                                                                                                            

In [17]: abs(x-2).rewrite(Piecewise)                                                                                                           
Out[17]: 
⎧x - 2  for x - 2 ≥ 0
⎨                    
⎩2 - x    otherwise 
```
We can get the boundaries of the condition from there
```
In [23]: abs(x-2).rewrite(Piecewise).args[0][1].as_set()                                                                                       
Out[23]: [2, ∞)
```
It would be useful to have a function that gives the points at which a complex function is not analytic.

mijo2 commented:
@oscarbenjamin For making such a function, say do we have ```f(x)``` then do we get the points ```solve(1/f(x))``` as the points that we want.

oscarbenjamin commented:
> get the points `solve(1/f(x))`

That is what singularities basically does although it could be extended to handle other kinds of functions.

The problem here is that we aren't looking for singularities. We're trying to find the non-differentiable points in a function like `abs(x)`.

mijo2 commented:
@oscarbenjamin So, for function like ```abs(x)``` we can get suspect points for example:

```
abs(x) = { x; x >= 0
              { -x; x < 0
```

If we can write a function in terms of continuous functions then we can get the points where the function maybe discontinuous or non-differentiable.

Also, to find the points where the function has infinite value of the derivative, we simply have to find the points where ```f'(x)**-1 = 0```. Will that give us the solutions that we require?

czgdp1807 commented:
Does the approach work with multivariate functions?

mijo2 commented:
@czgdp1807 I am not sure if it works. Will have to check and get back to you.

czgdp1807 commented:
If it doesn't work on multivariate variables then IMO, proper `NotImplementedError` should be raised in such cases for communicating the same to the end user.
Moreover, can you please explain the strategy you are using to compute the results?

mijo2 commented:
@czgdp1807 The argmax and argmin functions are for univariate and these points are calculated by taking the first order derivative and equating it to 0. Also, this uses the ```function_range``` function to get the supremums and the infimums and with both, the sups, infs and critical points(solutions of the 1st order DE), we can check which of the points constitute to argmax and argmin respectively. Also, a ```ValueError``` is raised when ```argmax``` and ```argmin``` for more than two or more free variables is tried to compute. Also, this approach works only when the equation is differentiable in the domain specified and the domain must be a subset of the Real domain.

czgdp1807 commented:
Can you show the output of your code for the function, `|x|`?

mijo2 commented:
@czgdp1807 The output is a ```ValueError```:

```
ValueError: 
Can't calculate derivative wrt 0.
```

czgdp1807 commented:
I would like to see, `0` as the output for `argmin`.

czgdp1807 commented:
See, https://en.wikipedia.org/wiki/Critical_point_(mathematics)#Application_to_optimization

mijo2 commented:
@czgdp1807 I will take a look at this and give an update soon.

mijo2 commented:
@czgdp1807 Regarding this PR, for now I am trying to implement a recursive check for getting critical points for a function.

oscarbenjamin commented:
A function to find the critical points would be useful for other things as well.

mijo2 commented:
@czgdp1807 Update for this PR, I haven't yet completed the function, but it rough looks something like below:

```
In [36]: def return_piecewise_solutions(f, x): 
    ...:     from sympy.logic.boolalg import And, Or 
    ...:      
    ...:     solutions = [] 
    ...:     if type(f) == Piecewise: 
    ...:         for arg in f.args: 
    ...:             values = [] 
    ...:             if isinstance(arg[1], (And, Or)): 
    ...:                 for a in arg[1].args: 
    ...:                     values += a.args 
    ...:             else: 
    ...:                 values = arg[1].args 
    ...:             for v in values: 
    ...:                 if v.is_Number: 
    ...:                     solutions.append(v) 
    ...:     return solutions
```

Here, for ```abs(x).rewrite(Piecewise)``` it gives ```[0]``` and for ```Piecewise((x, (x >= 10) | (x <= -10)), (x-1, True))``` it gives  ```[-10, 10]```. I have yet to make it recursive but if there are any problems with this kind of implementation then please suggest if need be and if possible.

czgdp1807 commented:
So, it looks like that your function will be able to handle any input function which contains `Piecewise`. IMO, the problem can be solved by implementing some optimisation algorithms. We can start with optimization of univariate functions. Can you find some algorithms in that area? Wolfram provides, `ArgMin` and `ArgMax`.
Here are some short notes,
```python
For linear cases, Minimize and Maximize use exact linear programming methods. For polynomial cases they use cylindrical algebraic decomposition. For analytic cases in closed and bounded boxes they use methods based on Lagrange multipliers.
For bounded linear cases with integer variables, Minimize and Maximize use a branch-and-bound method for integer linear programming, with preprocessing done by lattice reduction.
Piecewise functions are symbolically expanded, with each case being treated as a separate optimization problem.
```

mijo2 commented:
@czgdp1807 Sorry, was busy with some other PR, I will surely try to look for the methods that you described and start working on them soon. For now, the critical points extractor that I am implementing isn't complete and after it is complete, I will start working on this soon.

mijo2 committed:
Merge branch 'master' into 13432_argmax_min

mijo2 committed:
calculus: Added piecewise function in util.py

Added piecewise function extractor that extracts suspected critical
points from a function.

mijo2 commented:
I have added the recursive function that extracts the critical points. But now, this function still needs to be extended to handle more cases and finally change the exception around ```frange``` in ```_argMaxMin``` function to handle functions that raise a ```ValueError``` function.

oscarbenjamin commented on sympy/calculus/util.py:
```diff
@@ -1631,3 +1632,178 @@ def union(self, other):
+    f = _sympify(f)
+    if symbol is None:
+        if len(f.free_symbols) > 1:
+            raise ValueError(filldedent('''
```
I guess this should be `len(free_symbols) != 1`.

Maybe this logic could be moved to `_argMaxMin` if it otherwise needs to be duplicated in `argmax_real` and `argmin_real`

oscarbenjamin commented on sympy/calculus/util.py:
```diff
@@ -1631,3 +1632,178 @@ def union(self, other):
+    # If the supremum or the infimum of the function is not in its range,
+    # then an empty set is returned.
+    if not FiniteSet(extremum).is_subset(frange):
+        return EmptySet()
```
The `is_subset` function can return True, False or None (for indeterminate). We should only return the EmptySet if we are sure that extremum is not in frange.

It isn't necessary to create a FiniteSet when we can use `frange.contains(extremum)`. That also can return None (an unevaluated Contains) though.

oscarbenjamin commented on sympy/calculus/util.py:
```diff
@@ -1631,3 +1632,178 @@ def union(self, other):
+        soln1 = FiniteSet(domain.sup)
+        soln2 = FiniteSet(domain.inf)
+        if max:
+            return FiniteSet(list(soln1)[0]) if soln1.is_subset(domain) else S.EmptySet
```
How about:
```python
if max:
    point = domain.sup
else:
    point = domain.inf
inrange = domain.contains(point)
if inrange is S.true:
    return FiniteSet(point)
elif inrange is S.false:
    return S.EmptySet
else:
    # This is the None case. What do we do here?
```

mijo2 replied:
I don't understand when will contains give ```None``` output?

oscarbenjamin replied:
The `contains` method does not give None but in the indeterminate case it won't evaluate to True/False e.g.:
```julia
In [4]: FiniteSet(x, y).contains(z)                                                                                                                           
Out[4]: z ∈ {x, y}

In [5]: type(_)                                                                                                                                               
Out[5]: Contains
```

mijo2 replied:
Then we can raise a value error or a not implemented error I guess but i am not certain.

oscarbenjamin replied:
If the set is not there we return an empty set. If it is we return the finite set. If we aren't sure we can return an intersection. In fact we can just return an intersection and let that evaluate or not:
```julia
In [14]: FiniteSet(0) & Interval(1, 2)                                                                                                                        
Out[14]: ∅

In [15]: FiniteSet(1) & Interval(1, 2)                                                                                                                        
Out[15]: {1}

In [16]: FiniteSet(x) & Interval(1, 2)                                                                                                                        
Out[16]: [1, 2] ∩ {x}

In [17]: _.subs(x, 0)                                                                                                                                         
Out[17]: ∅
```

mijo2 replied:
@oscarbenjamin sure thing will do this.

oscarbenjamin commented:
There needs to be consideration of the indeterminate case where there are unknown symbols e.g.:
```julia
In [10]: from sympy.calculus.util import argmin_real                                                                                                          

In [11]: a, b, c, x = symbols('a, b, c, x', real=True)                                                                                                        

In [12]: argmin_real((x - a)**2, x, Interval(b, c))                                                                                                           
---------------------------------------------------------------------------
TypeError 
```
What should be returned in that case?

Currently it fails because `function_range` fails.

Perhaps the answer could be:
```julia
In [13]: FiniteSet(a) & Interval(b, c)                                                                                                                        
Out[13]: {a} ∩ [b, c]
```

mijo2 commented:
@oscarbenjamin You are correct about the limitations of ```frange``` function as of now. I will see what can be done to improve the function to treat other symbols as constants.

oscarbenjamin commented on sympy/calculus/util.py:
```diff
@@ -1631,3 +1632,168 @@ def union(self, other):
+        elif inrange is S.false:
+            return S.EmptySet
+        else:
+            return domain & FiniteSet(point)
```
I think we can just return this instead of having the if/elif above

oscarbenjamin commented on sympy/calculus/util.py:
```diff
@@ -1631,3 +1632,168 @@ def union(self, other):
+    """
+    f = _sympify(f)
+
+    return _argMaxMin(f, symbol , domain, max = False)
```
I guess the `_sympify` can be moved to `_argMaxMin` as well

oscarbenjamin commented:
> The following commits **add new files**:
> 
> * [110954b](https://github.com/sympy/sympy/commit/110954bd8801f0a5167dad3fdae5907117695225):
>   
>   * `piecewise_update`

I don't think that you meant to add this file

mijo2 commented:
@oscarbenjamin Yes I would delete that file in the next commit.

oscarbenjamin commented:
I would prefer it if that file is not in the commit history at all.

mijo2 commented:
Ok, @oscarbenjamin I would squash that commit.

oscarbenjamin commented on sympy/calculus/util.py:
```diff
@@ -1631,3 +1632,161 @@ def union(self, other):
+    from sympy.logic.boolalg import And, Or
+
+    solutions = []
+    if type(f) == Piecewise:
```
Maybe it would be better to use `piecewise_fold`

mijo2 replied:
@oscarbenjamin Yes that would be great. I didn't know this type of function existed. I would update the logic of this function that returns the suspected critical points in my next commit.

mijo2 committed:
calculus: Added some functionality to deal with abs(x)

Added some functionality which lets the _argMaxMin function deal
with functions that are discontinuous at finite number of points.
Still, this functionality can handle only one level of discontinuity
but that can be improved.

mijo2 commented:
@oscarbenjamin @czgdp1807 I haven't yet updated the test cases as I am still seeing if this works but I have added some code to deal with functions like ```abs(x)```. Please take a look and review if necessary.

mijo2 committed:
calculus: Updated return_piecewise_solutions function

Updated the function and added some test cases. There is still quite
some work left.

czgdp1807 commented on sympy/calculus/tests/test_util.py:
```diff
@@ -534,5 +534,48 @@ def test_issue_16469():
+    assert argmin_real(-x**2 + 4, x, Interval(-2, 2)) == FiniteSet(-2, 2)
+    assert argmin_real(exp(y)) == S.EmptySet
+    assert argmin_real(abs(x), x, S.Reals) == FiniteSet(0)
+    assert argmin_real(abs(2*x-1), domain=S.Reals) == FiniteSet(1/2)
```
Any test cases for domains like, `{1, 2, 3}` i.e., those which are finite or countably-infinite?

czgdp1807 commented on sympy/calculus/util.py:
```diff
@@ -1631,3 +1632,210 @@ def union(self, other):
 AccumBounds = AccumulationBounds
+
+
+def argmax_real(f, symbol=None, domain=S.Reals):
```
`argmax` sounds better. Sets like, `N` aren't perfectly real. For sets which are not subsets of real a `NotImplementedError` can be raised.

czgdp1807 commented on sympy/calculus/util.py:
```diff
@@ -1631,3 +1632,210 @@ def union(self, other):
+               if v.is_Number:
+                   solutions.append(v)
+
+    elif type(f) == Add or type(f) == Mul or type(f) == Pow:
```
`isinstance`?

czgdp1807 commented:
IMHO, algorithms in https://github.com/sympy/sympy/pull/18550#issuecomment-590475991 should be implemented for `argmin` and `argmax` which are in itself separate projects. I don't know if you are using the approach given in the above comment. I am for now attaching a `Could Close` label, though if you are using the approaches(used by wolfram, as provided in the above comment) then let me know, will remove the label.

czgdp1807 commented:
I also remember there was a requirement for adding a utility to compute critical points of a function. I think a separate issue can be opened for the same.

oscarbenjamin commented:
There are further improvements that can be made but perhaps they are for other PRs. In that case this PR looks reasonable to me. Should the `argmin_real` or `argmax_real` functions be added to any of the `.rst` docs?

mijo2 commented:
> IMHO, algorithms in #18550 (comment) should be implemented for argmin and argmax which are in itself separate projects. I don't know if you are using the approach given in the above comment.

You are correct, this PR would become too big if I add the approach you suggested in this one directly. 

> I also remember there was a requirement for adding a utility to compute critical points of a function. I think a separate issue can be opened for the same.

Yes, it is much a needed functionality not only for argmax and argmin but I am sure it can be helpful for many situations and solvers.

> Should the argmin_real or argmax_real functions be added to any of the .rst docs?

I can update the documentation to explain well how these functions operate, especially their limitations. The functionality I tried to add for Critical points is far from complete and there are many things to be taken care of. For now, it can handle only basic things like ```abs(x)``` among discontinuous functions and I am sure there are many cases of functions where this functionality would break. As of your suggestion for adding feature where ```abs(x - a)``` can be calculated, I was not able to accommodate it in this PR.

mijo2 committed:
calculus: Revived PR to add argmax_real and argmin_real functions

The PR that was supposed to start a series of fixes for #13432
was left as open for a few years because the author(me; mijo2)
was pulled in some other priority work. This commit revives the PR
by merging the new changes. There have been no big changes in
the implementation. But the unit tests had to be updated, particularly
for the equality of ImageSet values, which is done in a hacky but safe way
for now.

mijo2 committed:
calculus: Updated the PR to fix issue #13432 with new changes

When I was reviving the PR for the fix of issue #13432 in my local,
the master was updated. This commit pulls in the latest changes of
the master branch and also adds mijo2(me) with my new email in .mailmap
file. A small fix is also made to return S.EmptySet in util.py instead
of EmptySet()
